### DISCUSSION TITLE
What are the requirements to display release notes for a Gradle plugin?

### DISCUSSION THREAD

**Discussion by monosoul** [renovatebot/renovate#19514] [Request Help]
Hi all,

I maintain [this Gradle plugin](https://plugins.gradle.org/plugin/dev.monosoul.jooq-docker) and noticed that for some reason renovate bot doesn't fetch release notes, nor does it create a link to github for it.

Here's a repo reproducing the issue: https://github.com/monosoul/renovate-release-notes-issue-reproducer
There are 2 PRs there:

- One for `kotlin-jvm` plugin and it has release notes and a link to the sources: monosoul/renovate-release-notes-issue-reproducer#2
- One for `dev.monosoul.jooq-docker` plugin and it **doesn't have** release notes and a link to the sources: monosoul/renovate-release-notes-issue-reproducer#1

The plugin I maintain has a link to GitHub configured as both `vcsUrl` and `website`: https://github.com/monosoul/jooq-gradle-plugin/blob/main/buildSrc/src/main/kotlin/publishing-convention.gradle.kts#L17
It also has releases with notes for every release: https://github.com/monosoul/jooq-gradle-plugin/releases

So, what else is missing there? Why does renovate bot fails to show release notes for the plugin?

Thanks!

**Comment by viceice** [comment]
check the published XML / pom files. they maybe miss the urls?

**Reply by monosoul** [reply]
Thanks for the prompt reply! Indeed, they miss the urls. I expected gradle publish plugin to add them to the pom files :thinking: I'll have a look into that, thanks!

**Reply by monosoul** [reply]
@viceice I've tested it with [another plugin](https://plugins.gradle.org/plugin/com.github.monosoul.markdown.page.generator). The new version of this plugin [has all the links in it's `pom.xml`](https://plugins.gradle.org/m2/gradle/plugin/com/github/monosoul/markdown-page-generator-gradle-plugin/2.4.0.0/markdown-page-generator-gradle-plugin-2.4.0.0.pom), yet renovate still doesn't show release notes for it: https://github.com/monosoul/renovate-release-notes-issue-reproducer/pull/4

**Reply by monosoul** [reply]
Or will renovate use [this `pom.xml`](https://plugins.gradle.org/m2/com/github/monosoul/markdown/page/generator/com.github.monosoul.markdown.page.generator.gradle.plugin/2.4.0.0/com.github.monosoul.markdown.page.generator.gradle.plugin-2.4.0.0.pom) instead? If yes, then I think there's nothing I can do to add the urls there. I think Gradle's plugin portal trims it to to only have the minimum set of fields.

**Reply by viceice** [reply]
check the logs, you should see all urls there

here we construct the pom url for metadata:
https://github.com/renovatebot/renovate/blob/52d773bb6b5b62588fc6cfbb7750517899789b14/lib/modules/datasource/maven/util.ts#L383-L390

**Reply by monosoul** [reply]
@viceice  Looking at [the logs](https://app.renovatebot.com/dashboard#github/monosoul/renovate-release-notes-issue-reproducer/940732717) I think I see the issue there. Renovate assumes that the plugin is published to `repo.maven.apache.org`. At least that's where it gets `kotlin.jvm` plugin's metadata from. It makes 2 calls:
   - `https://plugins.gradle.org/m2/org/jetbrains/kotlin/jvm/org.jetbrains.kotlin.jvm.gradle.plugin/1.7.22/org.jetbrains.kotlin.jvm.gradle.plugin-1.7.22.pom`
   - `https://repo.maven.apache.org/maven2/org/jetbrains/kotlin/jvm/org.jetbrains.kotlin.jvm.gradle.plugin/1.7.22/org.jetbrains.kotlin.jvm.gradle.plugin-1.7.22.pom`

The first one doesn't have urls (since plugin portal trims the plugin coordinates pom file), the second one does. The problem is that when you publish a Gradle plugin - it creates 2 `pom.xml` files. The first one will be used as the plugin coordinates and will have a single dependency referring the actual jar file. The second pom will belong to the jar artifact itself and have all the data in it.

So if we take the Kotlin jvm plugin as an example, it will have 2 poms in Gradle's repo:

1. Plugin coordinates: `https://plugins.gradle.org/m2/org/jetbrains/kotlin/jvm/org.jetbrains.kotlin.jvm.gradle.plugin/1.7.22/org.jetbrains.kotlin.jvm.gradle.plugin-1.7.22.pom`
2. The actual artifact with all the metadata: `https://plugins.gradle.org/m2/org/jetbrains/kotlin/kotlin-gradle-plugin/1.7.22/kotlin-gradle-plugin-1.7.22.pom`

So currently renovate will only fetch release notes for Gradle plugins those are published to `repo.maven.apache.org` in addition to Gradle plugin portal.

**Reply by monosoul** [reply]
Probably renovate needs to have metadata resolution logic specific to Gradle plugins

**Reply by viceice** [reply]
@zharinov Any idea who could help here? I don't know the complex maven datasource very well.

@monosoul I can't see your app dashborad until i'm a maintainer of that repo. only @rarkins has access to internal logs.

**Reply by monosoul** [reply]
@viceice I've added you as a collaborator there, but not sure if that's enough. Here's the log: https://pastebin.com/UFTj2Vm2

**Comment by monosoul** [comment]
I guess depending on how you look at it, one might say that's an issue with renovate (since it's unaware of Gradle plugin portal's nuances), another might say it's an issue with Gradle plugin portal, since trimming the pom file is an unexpected behavior. So to make it fair, I have also raised an issue with Gradle plugin portal here: https://github.com/gradle/plugin-portal-requests/issues/173 .

But if this could be fixed without a huge effort on your end, I'd appreciate if you do that, since with Gradle it might take a while.
